File: my-solutions/0033-search-in-rotated-sorted-array/solution.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
      def binSearch(l, r, target):
        while(l <= r):
          mid = int((l + r) / 2)
          if nums[mid] == target:
            return mid
          elif nums[mid] > target:
            r = mid - 1
          else:
            l = mid + 1
        return -1
      
      if nums[len(nums) - 1] > nums[0]:
        return binSearch(0, len(nums) - 1, target)

      def searchTarget(l, r, target):
        if r - l <= 2:
          return nums.index(target) if target in nums else -1

        mid = int((l + r)/2)

        print(l, mid, r)

        if nums[mid] == target:
          return mid
        if nums[mid] < nums[l]:
          postMid = binSearch(mid + 1, r, target)
          return postMid if postMid != -1 else searchTarget(l, mid - 1, target)
        elif nums[mid] > nums[l]:
          preMid = binSearch(l, mid - 1, target)
          return preMid if preMid != -1 else searchTarget(mid + 1, r, target)

      return searchTarget(0, len(nums) - 1, target)

File: my-solutions/0033-search-in-rotated-sorted-array/solution.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // check nums[m] vs nums[l]
        // figure out which side of m is sorted (left or right)
        // move the l or r pointer depending on which side is sorted. We can only be sure of the side that is sorted.

        int l = 0, r = nums.size() - 1;
        while(l <= r) {
            int m = l + (r - l) / 2;
            if(nums[m] == target) return m;

            if(nums[m] >= nums[l]) {
                if(nums[m] > target && target >= nums[l]) r = m - 1;
                else l = m + 1;
            } else {
                if(nums[m] < target && target <= nums[r]) l = m + 1;
                else r = m - 1;
            }
        }

        return -1;
    }
};
